urlcdd0c1ab49_emiln_JHouseFireExplorer_tgz-pJ8-dk_brics_automaton_DatatypesJ8/src/dk/brics/automaton/SpecialOperations.java

```diff
@@ -0,0 +1,602 @@
+	 */
+	public static Automaton subst(Automaton a, char c, String s) {
+		a = a.cloneExpandedIfRequired();
+		Set<StatePair> epsilons = new HashSet<StatePair>();
+		for (State p : a.getStates()) {
+			Set<Transition> st = p.transitions;
+			p.resetTransitions();
+			for (Transition t : st)
+				if (t.max < c || t.min > c)
+					p.transitions.add(t);
+				else {
+					if (t.min < c)
+						p.transitions.add(new Transition(t.min, (char)(c - 1), t.to));
+					if (t.max > c)
+						p.transitions.add(new Transition((char)(c + 1), t.max, t.to));
+					if (s.length() == 0)
+						epsilons.add(new StatePair(p, t.to));
+					else {
+						State q = p;
+						for (int i = 0; i < s.length(); i++) {
+							State r;
+							if (i + 1 == s.length())
+								r = t.to;
+							else
+								r = new State();
+							q.transitions.add(new Transition(s.charAt(i), r));
+							q = r;
+						}
+					}
+				}
+		}
+		a.addEpsilons(epsilons);
+		a.deterministic = false;
+		a.removeDeadTransitions();
+		a.checkMinimizeAlways();
+		return a;
+	}
+	
+	/**
+	 * Returns an automaton accepting the homomorphic image of the given automaton
+	 * using the given function.
+	 * <p>
+	 * This method maps each transition label to a new value.
+	 * <code>source</code> and <code>dest</code> are assumed to be arrays of
+	 * same length, and <code>source</code> must be sorted in increasing order
+	 * and contain no duplicates. <code>source</code> defines the starting
+	 * points of char intervals, and the corresponding entries in
+	 * <code>dest</code> define the starting points of corresponding new
+	 * intervals.
+	 */
+	public static Automaton homomorph(Automaton a, char[] source, char[] dest) {
+		a = a.cloneExpandedIfRequired();
+		for (State s : a.getStates()) {
+			Set<Transition> st = s.transitions;
+			s.resetTransitions();
+			for (Transition t : st) {
+				int min = t.min;
+				while (min <= t.max) {
+					int n = findIndex((char)min, source);
+					char nmin = (char)(dest[n] + min - source[n]);
+					int end = (n + 1 == source.length) ? Character.MAX_VALUE : source[n + 1] - 1;
+					int length;
+					if (end < t.max)
+						length = end + 1 - min;
+					else
+						length = t.max + 1 - min;
+					s.transitions.add(new Transition(nmin, (char)(nmin + length - 1), t.to));
+					min += length;
+				}
+			}
+		}
+		a.deterministic = false;
+		a.removeDeadTransitions();
+		a.checkMinimizeAlways();
+		return a;
+	}
+	
+	/**
+	 * Returns an automaton with projected alphabet. The new automaton accepts
+	 * all strings that are projections of strings accepted by the given automaton
+	 * onto the given characters (represented by <code>Character</code>). If
+	 * <code>null</code> is in the set, it abbreviates the intervals
+	 * u0000-uDFFF and uF900-uFFFF (i.e., the non-private code points). It is
+	 * assumed that all other characters from <code>chars</code> are in the
+	 * interval uE000-uF8FF.
+	 */
+	public static Automaton projectChars(Automaton a, Set<Character> chars) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 28. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.